### PR TITLE
add unit tests to pipelinerun_helper

### PR DESCRIPTION
**Pull request by avi-biton**
Add unit tests to pipelinerun_helper

### CODE REVIEW

**Review by yftacherzog — COMMENTED** [review]
How does it look when one function is failing?

I was expecting the unit tests to be separated by the units (functions) being tested.

I do see some sense now with grouping them by context, which can represent a single scenario in which the different functions are called. having said that, I'd like to see that when we have a regression in one function, the results will point to that function, rather than to a case with many functions.

File: internal/controller/pipelinerun_helper_test.go [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+	v1 "knative.dev/pkg/apis/duck/v1"
+)
+
+var _ = Describe("Unit testing for pipelinerun_helper", func() {
+	var (
+		testTruePipelineRun, testFalsePipelineRun, testErrMmarshelPipelineRun *tektonv1.PipelineRun
+		notPipelineRun                                                        *tektonv1.TaskRun
+	)
+
+	Describe("Push Pipelinerun Ended with finalizer, annotation and results", func() {
```
Comment (yftacherzog) [inline internal/controller/pipelinerun_helper_test.go:21]: Let's add some commas/semi-colons or something to make sure the description is read in the way we intended to (same for other test cases as well).
```suggestion
	Describe("Push Pipelinerun; ended; with finalizer, annotation and results", func() {
```

Comment (avi-biton) [reply internal/controller/pipelinerun_helper_test.go]: Done

**Comment by avi-biton** [comment]
> How does it look when one function is failing?
> 
> I was expecting the unit tests to be separated by the units (functions) being tested.
> 
> I do see some sense now with grouping them by context, which can represent a single scenario in which the different functions are called. having said that, I'd like to see that when we have a regression in one function, the results will point to that function, rather than to a case with many functions.

It is well organized when using the `Context` and `It` structure.
I failed one test so you can see how it looks.

File: internal/controller/pipelinerun_helper_test.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+							Message: "Tasks Completed: 12 (Failed: 0, Cancelled 0), Skipped: 2",
+							Reason:  "Completed",
+							Status:  "True",
+							Type:    apis.ConditionSucceeded,
+						},
+					},
+				},
+			},
+		}
+		Context("when a push pipelinerun is created, end successfully and has results", func() {
```
Comment (yftacherzog) [inline internal/controller/pipelinerun_helper_test.go:81]: We might not need the context, as I'm not sure it adds any new information, but if we keep it, I think it should say something like this:
```suggestion
		Context("when a push pipelinerun ends successfully and has results", func() {
```

File: internal/controller/pipelinerun_helper_test.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+			It("should be a push pipelinerun", func() {
```
Comment (yftacherzog) [inline internal/controller/pipelinerun_helper_test.go:82]: The descriptions still confuse me.
We made it a push pipeline. We're now testing that our function indicates the same.
I think it should be something like this (with similar changes to some of the other calls).
```suggestion
			It("IsPushPipelineRun should return true", func() {
```

**Comment by yftacherzog** [comment]
> > How does it look when one function is failing?
> > I was expecting the unit tests to be separated by the units (functions) being tested.
> > I do see some sense now with grouping them by context, which can represent a single scenario in which the different functions are called. having said that, I'd like to see that when we have a regression in one function, the results will point to that function, rather than to a case with many functions.
> 
> It is well organized when using the `Context` and `It` structure. I failed one test so you can see how it looks.

ok. I'm good with that. In that case, I think the `It`s should mention the name of the function being tested (left a comment with an example).

File: internal/controller/pipelinerun_helper_test.go (hunk 2 above)
Comment (avi-biton) [reply internal/controller/pipelinerun_helper_test.go]: Done

File: internal/controller/pipelinerun_helper_test.go (hunk 3 above)
Comment (avi-biton) [reply internal/controller/pipelinerun_helper_test.go]: Done

File: internal/controller/pipelinerun_helper_test.go [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+			It("IsPipelineRunEndedSuccessfully should return false", func() {
+				got := IsPipelineRunEndedSuccessfully(notPipelineRun)
+				Expect(got).To(BeFalse())
+			})
+		})
+	})
+	Describe("When results fails to marshal", func() {
+		testErrMmarshelPipelineRun = &tektonv1.PipelineRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pull_request",
```
Comment (yftacherzog) [inline internal/controller/pipelinerun_helper_test.go:263]: The actual type doesn't fit the name.

File: internal/controller/pipelinerun_helper_test.go [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+				Expect(err).To(HaveOccurred())
+				Expect(err.Error()).To(ContainSubstring("failed to marshel results from pipelinerun %s", testErrMmarshelPipelineRun.Name))
+			})
+		})
+	})
+
+	Describe("Successfully adding/removing resources when not exist", func() {
+		BeforeEach(func() {
+			// Create a pull_request pipelinerun
+			prPipelineRun = &tektonv1.PipelineRun{
```
Comment (yftacherzog) [inline internal/controller/pipelinerun_helper_test.go:315]: Does it have a type at all? The name might need to change.

File: internal/controller/pipelinerun_helper_test.go [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+		})
+		AfterEach(func() {
+			err := k8sClient.Delete(ctx, testPipelineRun)
+			Expect(err == nil || errors.IsNotFound(err)).To(BeTrue())
+		})
+	})
+	Describe("Successfully adding/removing resources when exist", func() {
+		BeforeEach(func() {
+			// Create a pull_request pipelinerun
+			prPipelineRun = &tektonv1.PipelineRun{
```
Comment (yftacherzog) [inline internal/controller/pipelinerun_helper_test.go:396]: type-name matching

**Review by yftacherzog — COMMENTED** [review]
Left some minor comments

File: internal/controller/pipelinerun_helper_test.go (hunk 5 above)
Comment (avi-biton) [reply internal/controller/pipelinerun_helper_test.go:315]: In line 24:
`prPipelineRun *tektonv1.PipelineRun`

File: internal/controller/pipelinerun_helper_test.go (hunk 6 above)
Comment (avi-biton) [reply internal/controller/pipelinerun_helper_test.go:396]: In line 24:
`prPipelineRun *tektonv1.PipelineRun`

File: internal/controller/pipelinerun_helper_test.go (hunk 4 above)
Comment (avi-biton) [reply internal/controller/pipelinerun_helper_test.go:263]: fixed

File: internal/controller/pipelinerun_helper_test.go (hunk 5 above)
Comment (yftacherzog) [reply internal/controller/pipelinerun_helper_test.go:315]: But here we assign to that variable and I don't think there's anything that actually makes it a pr pipelinerun other than the name.

Comment (avi-biton) [reply internal/controller/pipelinerun_helper_test.go:315]: You are right, I have removed the pull_request label, let me fix it

**Pull request merged**